### PR TITLE
Don't use configured CFLAGS & CPPFLAGS to compile third-party C sources

### LINKED ISSUE #12578: Should the compiler's compile-time CFLAGS be hardcoded into ocamlc/ocamlopt?
Hi,

I have been running into some unexpected behavavior with the default C compiler flags of ocamlc/ocamlopt on Cygwin64 with OCaml 4.14.0.

See some initial discussion here: https://cygwin.com/pipermail/cygwin-apps/2023-August/043106.html 

When `ocamlc` compiles C programs, it gets passed some options
```
$ ocamlc hello.c -verbose
+ gcc -O2 -fno-strict-aliasing -fwrapv -pthread  -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -fdebug-prefix-map=some_build_path1=/usr/src/debug/ocaml-4.14.0-1 -fdebug-prefix-map=some_build_path2=/usr/src/debug/ocaml-4.14.0-1 -D_FILE_OFFSET_BITS=64 -U_WIN32    -c    -I'/usr/lib/ocaml' 'hello.c'
```
which are a superset of the package-time `CFLAGS` used to build the OCaml compiler itself. The package-time `CFLAGS` are hardcoded into the `ocamlc` executable itself.

Reading https://github.com/ocaml/ocaml/pull/9824, this appears to be an intentional design choice on OCaml's part but it can still cause odd behavior.

In short, all end user C programs compiled with ocamlc will inherit the options that were used to compile the compiler (that could range from very helpful to rather unnecessary options for most cases). However, it's possible some end user programs might not want some of the package-time options (`-O2`, `-Wall`, `-Werror`, `-fdebug-prefix-map`, etc.)

It is possible to negate some of these options on the user side; for example, if one wants to compile without debug info they need to remember to always call `ocamlc -ccopt -ggdb0`.

Even this might not always work: as far as I understand, GCC options with the property "RejectNegative" (https://gcc.gnu.org/onlinedocs/gccint/Option-properties.html) don't have a negative form and thus most of them can't be disabled in an obvious way. 

Previous proposals to include more flags in `ocamlopt` have been wary about defaults that are too strict; e.g. https://github.com/ocaml/ocaml/issues/12050

>"Enabling -ccopt "-Werror -Wall" when using ocamlc/ocamlopt to build C files is a choice of the build system, not something to be decided by the compiler itself."

However, the "compiler deciding itself" is effectively what happens when package time `CFLAGS` are hardcoded into `ocamlc`.

This places a silent but unexpected burden on users to inspect the `ocamlc` default compiler flags and disable ones they deem unnecessary. Is this phenomenon supposed to be part of the intended design choice?

### PR DESCRIPTION
**Pull request by shindere**
When C flags or C preprocessor flags are passed to configure, they
should be used to build the compiler itself only, not when the
compiler is invoked to compile third-party C sources, like stubs.

The way flags were computed for the Sun C compiler looked suspiscious,
so it gets updated here, too.

### CODE REVIEW

**Comment by gasche** [comment]
I am worried that existing packagers of OCaml for Linux distributions could rely on the current interface to pass hardening flags, so for them the change would be a regression. We should check with the Debian and Fedora/RH people how their current packages handle this. (cc @glondu @rwmjones).

**Comment by glondu** [comment]
> I am worried that existing packagers of OCaml for Linux distributions could rely on the current interface to pass hardening flags, so for them the change would be a regression.

With my Debian hat on: this reliance is accidental.

Instead, the rule is to use CFLAGS et al. at **package build time**. It is populated automatically by the Debian toolchain during package building, but users are free to set it to whatever they like for non-system software.

However, it turns out that these variables may be ignored by the build system, see for example https://github.com/ocaml/dune/pull/8464.

**Comment by rwmjones** [comment]
Adding @jamesjer too for Fedora

This change does seem totally backwards.  We want the hardening flags to be passed through when compiling 3rd party software (or a way for us to be able to specify them to be passed through).

It has been an ongoing problem that there is no way to pass specific flags for the compiler, and for what the compiler passes through, which we have attempted to solve with downstream patching in the past eg this commit (which we have since dropped):
https://pagure.io/fedora-ocaml/c/2501cd8b3c40188e5eb6dd2158d61cfa023236c2

**Comment by rwmjones** [comment]
Also this commit (also now dropped): https://pagure.io/fedora-ocaml/c/05b882737e0daecb86a6ef54e192508c60efcd88?branch=fedora-37-4.14.0

**Comment by shindere** [comment]
@rwmjones so AIUI there are two needs: (1) the need for a way _not_ to
have those CFLAGS used to compile OCaml propagate to the compilation of
third-party code and (2) (the need you expressed) to be able to specify,
at OCaml configuration time, C compiler flags that _will_ be used when
ocamlc or ocamlopt are used to compile third-party C sources. Is that a
correct sumary?

If so: I do think both needs can be fulfilled. The current  PR does not
fulfill them in its present form, and I think your needs can be
addressed in a  separate PR because they are distinct.

In terms of how to address them, the best way I can think would be to
add an `OCAMLC_CFLAGS` configuration variable where you could put the
flags you wnat to hardcode into the compiler. Would that seem a right
approach?

**Comment by rwmjones** [comment]
Yes, that seems better.  Jerry what do you think?

As background to this we need to run `annocheck` across both the OCaml compiler and OCaml programs.  This will check that hardening flags were used to compile every bit of code, including the OCaml runtime (`libasmrun.a` and friends) and C objects which are linked in and might have been compiled using `ocamlopt` as a wrapper.  Getting this right has occupied a lot of our time.

**Comment by gasche** [comment]
> The current  PR does not fulfill them in its present form, and I think your needs can be addressed in a  separate PR because they are distinct.

Maybe this does not need to be done in one go, but I would prefer to see a "better approach to pass C compilation flags to the OCaml compilers" merged before (or at the same time that) we remove the current mechanism that (some) distributions rely on.

**Comment by shindere** [comment]
Gabriel Scherer (2023/09/21 03:58 -0700):
> > The current  PR does not fulfill them in its present form, and I think your needs can be addressed in a  separate PR because they are distinct.
>
> Maybe this does not need to be done in one go, but I would prefer to
> see a "better approach to pass C compilation flags to the OCaml
> compilers" merged before (or at the same time that) we remove the
> current mechanism that (some) distributions rely on.

I can try. But first I need to make sure we do pass CFLAGS everywhere in
our own build system. On trunk we were propagating them too far, so to
speak, but now with the current implementaiotn we do'nt propagate them
far enough. That's the first thing to stamibise, I think. And I
uderstandthat, if this gets merged before the next release, then we will
also need  to merge something that makes it possible to pass flags to be
used by the compiler. Of course user programs always have the
possibility to pass options manually with `-ccopt` but I do understand
it's highly non-trivial to convince a user package's build system of
doing that.

**Comment by jamesjer** [comment]
I agree with @rwmjones 's comments.  I want to point out that we want to set both compiler and linker flags, including invocations of the linker from ocamlmklib.

**Comment by shindere** [comment]
I just pushed an updated version of this PR. The purpose of the PR remains the same, i.e. making sure that the C compiler/preprocessor flags passed to `configure` in the `CFLAGS`/`CPPFLAGS` configuration variables are used only during the build of the compiler and runtime themselves, but not when ocamlc or ocamlopt are used to compile third-party C sources (see issue #12578). The current version of the PR also takes into account both #13200 which reduced the scope of `CFLAGS` by not taking them into account at link time and #13201 which re-introduced the distinction between C compiler/preprocessor flags used to compile C codde for bytecode programs and those used to compile C code for native programs, as this distinction was not present when the present PR was initially submitted.

@rwmjones: this PR takes into account [your comment](https://github.com/ocaml/ocaml/pull/12589#issuecomment-1729159569) by introducing new configuration variables: `COMPILER_{BYTECODE,NATIVE}_{CFLAGS,CPPFLAGS}` which make it possible to specify at compiler configuration time which C compiler/preprocesor flags should be used when compiling C sources through ocamlc/ocamlopt to be linked with bytecode/native programs.

This new version of the PR thus superseeds #12975 by taking into account #13200 and #13201.

The revised implementation starts with a commit ensuring that the `-c` flag always appears at the end of the comilation rule as this one should, like `-o`, never be overriden.

The core of the work lives in the PR's second commit.

A third commit fixes the configuration for the Sun C compiler where the CFLAGS were dealt with whereas they should not have been.

Finally, this updated version of the PR fixes a duplication of flags related to pthread in the build system that has been reported by @tmcgilchrist (thanks!). It was causing no harm but was not very elegant either.

@jamesjer at this stage I am unsure how to address [your comment](https://github.com/ocaml/ocaml/pull/12589#issuecomment-1732040105). Is the current support of `LDFLAGS` not good enough and, if not, are you able to explain in more details what is missing?

File: Changes
Comment (ghost) [inline Changes]:
```suggestion
  C sources through the compiler. Flags for compiling third-party C
```

**Comment by shindere** [comment]
Fixed! Thanks!

**Comment by tmcgilchrist** [comment]
The duplication of `pthread` is fixed, thank you @shindere. One small nitpick, a few variables have extra whitespace padding that could be trimmed. For example:
```
BYTECCLIBS=-L/opt/homebrew/opt/zstd/lib -lzstd    -lpthread
BYTECODE_CFLAGS=-O2 -fno-strict-aliasing -fwrapv  -pthread 
NATIVECCLIBS=   -lpthread
```

**Comment by jamesjer** [comment]
The flags you added for @rwmjones sound like what I am interested in.  Thank you for that.  Are the flags passed to the linker by ocamlmklib affected, or just flags passed by ocamlc and ocamlopt?  I would like to have some way to set those passed by ocamlmklib, if possible.

**Comment by shindere** [comment]
Tim McGilchrist (2024/06/24 15:33 -0700):
> The duplication of `pthread` is fixed, thank you @shindere.

Thanks a lot for having double checked!

> One small nitpick, a few variables have extra whitespace padding that
> could be trimmed. [...]

Indeed. It turns out that avoiding superfluous spaces in flag lists is non-trivial as you must check, each time you add a flag to a list, whether the original list is empty (in whihc case the sesulting list is just that flag), or not (in which case the resulting list is the original one plus a space plus the flag).

Although we do make efforts to avoid superfluous spaces, this is not done systematically yet and we often add a space before the flag, without checking whether this is necessary or not, hence all these spaces you see.

One possible improvement consists in using the [AX_APPEND_FLAG](http://www.gnu.org/software/autoconf-archive/ax_append_flag.html) and [AX_PREPEND_FLAG](http://www.gnu.org/software/autoconf-archive/ax_prepend_flag.html) macros provided as part of the [GNU Autoconf archive](http://www.gnu.org/s/autoconf-archive/). These macros do not only avoid the superfluous spaces, they also check whether an added flag is already present in the variable to which it gets added to avoid repeating it.

I will be happy to submit a PR implementing this improvement, but would rather keep it distinct from the current one as the aims are different, and wait until this one has been merged to avoid the rebase hurdle, as the two PRs have code portions in common.

In any case, many thanks for the note @tmcgilchrist, which motivated me to look into this in more details.

For the sake of completeness, let me also say that, even if the varialbes themselves stop containing superfluous spaces, it may still be the case that the commands output by `make` will contain some, e.g. when certain variables are empty. It will thus be helpful at some point to make sure that all the commands output by make are first passed to its [strip](https://www.gnu.org/s/make/manual/html_node/Text-Functions.html) function so that they get normalized, which will make it easier to compare build logs by avoiding false positive due to changes in spacing.

PS: it's too bad GH does not interprete Markdown when replying by e-mail.

**Comment by rwmjones** [comment]
Whitespace is cosmetic and shouldn't make any difference, I wouldn't worry about it.

**Comment by shindere** [comment]
Jerry James (2024/06/24 19:56 -0700):
> The flags you added for @rwmjones sound like what I am interested in.
> Thank you for that.

Good. Thanks for confirming.

> Are the flags passed to the linker by ocamlmklib affected, or just
> flags passed by ocamlc and ocamlopt?

I am not sure I understand your question, since ocamlmklib is just a
wrapper. In the end, it does call ocamlc and ocamlopt.

> I would like to have some way to set those passed by ocamlmklib, if
> possible.

Are you able to be more precise here and give a concrete example of what
you need and that does not work yet?

FWIW, ocamlmklib can't be given C files as input. Also, it is just a
wrapper around ocamlopt and ocamlc so it ends up calling them.

For what I can tell, the flags put into the `LDFLAGS` configure variable
will be embedded in the compiler and used both internally (when we use
the compiler to compile our own tools) and externally, when the compiler
is used to compile (here link!) user programs.

**Comment by purplearmadillo77** [comment]
Thanks for the comprehensive work and the bugfix! AFAICT this successfully prevents CFLAGS options from being passed to the user. With this pull request on the latest trunk, `./configure CFLAGS=-Dfoo` on Cygwin produces
```
ocamlc hello.c -verbose
+ gcc -O2 -fno-strict-aliasing -fwrapv -pthread  -D_FILE_OFFSET_BITS=64 -U_WIN32   -c -I'/usr/lib/ocaml' 'hello.c'
```
as desired.

**Comment by shindere** [comment]
thanks a lot for having taken the time to checkthat the fix was actually
working and to reporting very clearly about your experiment.

**Review by dra27 — APPROVED** [review]
Good to go, thanks, @shindere!

**Comment by dra27** [comment]
I've put the PR [through precheck](https://ci.inria.fr/ocaml/job/precheck/979/) just to test the couple of more obscure platforms available there, but good to merge once Changes is updated and that has passed.

**Comment by shindere** [comment]
David Allsopp (2024/07/16 05:26 -0700):
> @dra27 approved this pull request.
>
> Good to go, thanks, @shindere!

Another bunch of thanks for your help, here again!

**Pull request merged**
